packages/core/src/provider/base.ts

```diff
@@ -508,36 +523,66 @@ export class SmartAccountProvider<
     return struct;
   };
 
-  readonly feeDataGetter: AccountMiddlewareFn = async (struct) => {
-    const [maxPriorityFeePerGas, feeData] = await Promise.all([
-      this.rpcClient.estimateMaxPriorityFeePerGas(),
-      this.rpcClient.estimateFeesPerGas(),
-    ]);
-    if (!feeData.maxFeePerGas || !feeData.maxPriorityFeePerGas) {
-      throw new Error(
-        "feeData is missing maxFeePerGas or maxPriorityFeePerGas"
+  readonly feeDataGetter: AccountMiddlewareFn = async (struct, overrides) => {
+    // maxFeePerGas must be at least the sum of maxPriorityFeePerGas and baseFee
+    // so we need to accommodate for the fee option applied maxPriorityFeePerGas for the maxFeePerGas
+    //
+    // Note that if maxFeePerGas is not at least the sum of maxPriorityFeePerGas and required baseFee
+    // after applying the fee options, then the transaction will fail
+    //
+    // Refer to https://docs.alchemy.com/docs/maxpriorityfeepergas-vs-maxfeepergas
+    // for more information about maxFeePerGas and maxPriorityFeePerGas
+    const estimateMaxFeePerGas = async (maxPriorityFeePerGas: BigNumberish) => {
+      const feeData = await this.rpcClient.estimateFeesPerGas();
+      if (!feeData.maxFeePerGas || !feeData.maxPriorityFeePerGas) {
+        throw new Error(
+          "feeData is missing maxFeePerGas or maxPriorityFeePerGas"
+        );
+      }
+
+      return (
+        BigInt(feeData.maxFeePerGas) -
+        BigInt(feeData.maxPriorityFeePerGas) +
+        BigInt(maxPriorityFeePerGas)
       );
-    }
+    };
 
-    // set maxPriorityFeePerGasBid to the max between 33% added priority fee estimate and
-    // the min priority fee per gas set for the provider
-    const maxPriorityFeePerGasBid = bigIntMax(
-      bigIntPercent(
-        maxPriorityFeePerGas,
-        BigInt(100 + this.maxPriorityFeePerGasEstimateBuffer)
-      ),
-      this.minPriorityFeePerBid
-    );
+    struct.maxPriorityFeePerGas =
+      overrides?.maxPriorityFeePerGas ??
+      (await this.rpcClient.estimateMaxPriorityFeePerGas());
+    struct.maxFeePerGas =
+      overrides?.maxFeePerGas ??
+      (await estimateMaxFeePerGas(struct.maxPriorityFeePerGas));
+
+    return struct;
+  };
 
-    const maxFeePerGasBid =
-      BigInt(feeData.maxFeePerGas) -
-      BigInt(feeData.maxPriorityFeePerGas) +
-      maxPriorityFeePerGasBid;
+  readonly feeOptionsMiddleware: AccountMiddlewareFn = async (
```

Comment:
fee options set on the provider were initially only meant to be used with the default gas estimator and default fee estimator. This was to allow someone to configure the bounds we use without having to write your own estimators. I'm not sure we need to support customizing them via a middleware though. IF someone overrides feeDataGetter or gasEstimator, then they would add their own custom logic and we shouldn't tightly couple overriding `feeDataGetter` and `feeOptionsMiddleware`. I think this creates a weird experience. EG:

Dev overrides feeDataGetter with their custom logic. Dev doesn't override feeOptionsMiddleware. Dev writes test for their override and expects values based on their implementation. Dev gets different values because the `feeOptionsMiddleware` affects those values